### PR TITLE
Qualimap

### PR DESCRIPTION
**Pull request by tverbeiren**
## Description

Added qualimap component for RNAseq QC

## Checklist before requesting a review

- [x] I have performed a self-review of my code

- [x] Conforms to the [Contributing guidelines](https://github.com/viash-hub/base/blob/main/CONTRIBUTING.md)

- [x] Proposed changes are described in the [CHANGELOG.md](https://github.com/viash-hub/base/blob/main/CHANGELOG.md)

- [x] I have tested my code with `viash ns test --parallel -q <name or namespace>`

- Check the correct box. Does this PR contain:
  - [ ] Breaking changes
  - [x] New functionality
  - [ ] Major changes
  - [ ] Minor changes
  - [ ] Documentation
  - [ ] Bug fixes

### CODE REVIEW

File: src/qualimap/qualimap_rnaseq/test.sh [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+echo ">> Checking output"
+[ ! -f "report.html" ] && echo "report.html does not exist" && exit 1
+[ ! -f "counts.txt" ] && echo "counts.txt does not exist" && exit 1
+[ ! -f "output.txt" ] && echo "output.txt does not exist" && exit 1
+
+echo ">> Checking if output is empty"
+[ ! -s "report.html" ] && echo "report.html is empty" && exit 1
+[ ! -s "counts.txt" ] && echo "counts.txt is empty" && exit 1
+[ ! -s "output.txt" ] && echo "output.txt is empty" && exit 1
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/test.sh:25]: Can we check whether the outputs make sense?

File: src/qualimap/qualimap_rnaseq/script.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+qualimap rnaseq \
+    --java-mem-size=$par_java_memory_size \
+    --algorithm $par_algorithm \
+    --num-pr-bases $par_pr_bases \
+    --num-tr-bias $par_tr_bias \
+    --sequencing-protocol $par_sequencing_protocol \
+    -bam $par_input \
+    -gtf $par_gtf \
+    -outdir "$tmp_dir" \
+    --java-mem-size=$par_java_memory_size \
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/script.sh:31]:
```suggestion
```

duplicate ^^

**Review by rcannood — CHANGES_REQUESTED** [review]

File: src/qualimap/qualimap_rnaseq/test.sh (hunk 1 above)
Comment (rcannood) [reply src/qualimap/qualimap_rnaseq/test.sh:25]: In the meantime, I added a bit more info to the contributing guidelines on what I mean by this: https://github.com/viash-hub/biobox/blob/main/CONTRIBUTING.md#step-12-create-test-script :relaxed:

**Comment by rcannood** [comment]
@dorien-er Any updates on this? :bow:

**Comment by dorien-er** [comment]
> @dorien-er Any updates on this? 🙇

Thx for the reminder, I somehow missed the review! But requested changes are made now.

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 3]
```diff
@@ -0,0 +1,111 @@
+name: "qualimap"
+info:
+  migration_info:
+    git_repo: https://github.com/nf-core/rnaseq.git
+    paths: [modules/nf-core/qualimap/rnaseq/main.nf]
+    last_sha: 54721c6946daf6d602d7069dc127deef9cbe6b33
+description: |
+  Qualimap RNA-seq QC RNA-seq QC reports quality control metrics and bias estimations 
+  which are specific for whole transcriptome sequencing, including reads genomic 
+  origin, junction analysis, transcript coverage and 5’-3’ bias computation.
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:10]: Could you fill in the missing metadata?

Also, the 'migration_info' can be omitted, because this component should be written from scratch

File: src/qualimap/qualimap_rnaseq/script.sh [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+# disable flags
+[[ "$par_paired" == "false" ]] && unset par_paired
+[[ "$par_sorted" == "false" ]] && unset par_sorted
+
+# Run qualimap
+qualimap rnaseq \
+    --java-mem-size=$par_java_memory_size \
+    --algorithm $par_algorithm \
+    --num-pr-bases $par_pr_bases \
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/script.sh:25]: Rename `--pr_bases` to `--num_pr_bases`

File: src/qualimap/qualimap_rnaseq/script.sh [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    --num-tr-bias $par_tr_bias \
+    --sequencing-protocol $par_sequencing_protocol \
+    -bam $par_input \
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/script.sh:28]: Rename `--input` to `--bam`

File: src/qualimap/qualimap_rnaseq/script.sh [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+# disable flags
+[[ "$par_paired" == "false" ]] && unset par_paired
+[[ "$par_sorted" == "false" ]] && unset par_sorted
+
+# Run qualimap
+qualimap rnaseq \
+    --java-mem-size=$par_java_memory_size \
+    --algorithm $par_algorithm \
+    --num-pr-bases $par_pr_bases \
+    --num-tr-bias $par_tr_bias \
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/script.sh:26]: rename `--tr_bias` to `--num_tr_bias`

File: src/qualimap/qualimap_rnaseq/script.sh [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    --sequencing-protocol $par_sequencing_protocol \
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/script.sh:27]: This argument is optional

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+  - name: "--paired"
+    type: boolean_true
+    description: Setting this flag for paired-end experiments will result in counting fragments instead of reads.
+  - name: "--sorted"
+    type: boolean_true
+    description: Setting this flag indicates that the input file is already sorted by name. If flag is not set, additional sorting by name will be performed. Only requiredfor paired-end analysis.
+  - name: "--java_memory_size"
+    type: string
+    required: false
+    default: 4G
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:81]: Don't set defaults, just add them to the docs.

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    type: string
+    required: false
+    choices: ["uniquely-mapped-reads", "proportional"]
+    default: uniquely-mapped-reads
+    description: Counting algorithm (uniquely-mapped-reads (default) or proportional).
+  - name: "--sequencing_protocol"
+    type: string
+    required: false
+    choices: ["non-strand-specific", "strand-specific-reverse", "strand-specific-forward"]
+    default: non-strand-specific
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:70]: Don't set defaults, just add them to the docs.

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    type: integer
+    required: false
+    default: 1000
+    min: 1
+    description: Number of top highly expressed transcripts to compute 5'-3' bias (default = 1000).
+  - name: "--algorithm"
+    type: string
+    required: false
+    choices: ["uniquely-mapped-reads", "proportional"]
+    default: uniquely-mapped-reads
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:64]: Don't set defaults, just add them to the docs.

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+  - name: "--pr_bases"
+    type: integer
+    required: false
+    default: 100
+    min: 1
+    description: Number of upstream/downstream nucleotide bases to compute 5'-3' bias (default = 100).
+  - name: "--tr_bias"
+    type: integer
+    required: false
+    default: 1000
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:57]: Don't set defaults, just add them to the docs.

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    required: false
+    example: report.html
+    description: Report output file. Supported formats are PDF or HTML.
+
+- name: "Optional"
+  arguments: 
+  - name: "--pr_bases"
+    type: integer
+    required: false
+    default: 100
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:51]: Don't set defaults, just add them to the docs.

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+resources:
+  - type: bash_script
+    path: script.sh
+test_resources:
+  - type: bash_script
+    path: test.sh
+  - path: test_data/
+
+engines:
+- type: docker
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:93]: Don't forget to add the version number to /var/software_versions.txt

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        cp -a qualimap_v2.3/. usr/bin && \
+        unset DISPLAY && \
+        mkdir -p tmp && \
+        export _JAVA_OPTIONS=-Djava.io.tmpdir=./tmp
+    - type: r
+      bioc: [ NOISeqr ]
+      cran: [ optparse ]
+runners: 
+- type: executable
+- type: nextflow
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:111]: Add newline at the end of this file

File: src/qualimap/qualimap_rnaseq/help.txt [hunk 15]
```diff
@@ -0,0 +1,52 @@
+QualiMap v.2.3
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/help.txt:1]: Since this component seems to be a port from nf-core, can you make sure that all arguments have been added to the component? Looks ok to me, but can you make sure?

**Review by rcannood — CHANGES_REQUESTED** [review]

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml (hunk 13 above)
Comment (dorien-er) [reply src/qualimap/qualimap_rnaseq/config.vsh.yaml:93]: hardcoded for now, cli tool doesn't seem to have param to fetch version

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+  which are specific for whole transcriptome sequencing, including reads genomic 
+  origin, junction analysis, transcript coverage and 5’-3’ bias computation.
+links:
+  homepage: http://qualimap.conesalab.org/
+  documentation: http://qualimap.conesalab.org/doc_html/analysis.html#rna-seq-qc
+  issue_tracker: https://bitbucket.org/kokonech/qualimap/issues?status=new&status=open
+  repository: https://bitbucket.org/kokonech/qualimap/commits/branch/master
+references:
+  doi: https://10.1093/bioinformatics/btv566
+license: GNU GPL v2
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:14]: see https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/licensing-a-repository

```suggestion
license: GPL-2.0
```

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+      type: file
+      required: false
+      direction: output
+      description: Output file for computed counts.
+    - name: "--output_report"
+      type: file
+      direction: output
+      required: false
+      example: report.html
+      description: Report output file. Supported formats are PDF or HTML.
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:50]: What would you think about:

```suggestion
    - name: "--qc_results"
      direction: output
      type: file
      required: true
      example: rnaseq_qc_results.txt
      description: Text file containing the RNAseq QC results.
    - name: "--counts"
      type: file
      required: false
      direction: output
      description: Output file for computed counts.
    - name: "--report"
      type: file
      direction: output
      required: false
      example: report.html
      description: Report output file. Supported formats are PDF or HTML.
```

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+    - name: "--num_tr_bias"
+      type: integer
+      required: false
+      min: 1
+      description: Number of top highly expressed transcripts to compute 5'-3' bias (default = 1000).
+    - name: "--algorithm"
+      type: string
+      required: false
+      choices: ["uniquely-mapped-reads", "proportional"]
+      default: uniquely-mapped-reads
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:68]: Remove default

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+      description: Counting algorithm (uniquely-mapped-reads (default) or proportional).
+    - name: "--sequencing_protocol"
+      type: string
+      required: false
+      choices: ["non-strand-specific", "strand-specific-reverse", "strand-specific-forward"]
+      default: non-strand-specific
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:74]: Remove default

File: src/qualimap/qualimap_rnaseq/config.vsh.yaml [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+    - name: "--paired"
+      type: boolean_true
+      description: Setting this flag for paired-end experiments will result in counting fragments instead of reads.
+    - name: "--sorted"
+      type: boolean_true
+      description: Setting this flag indicates that the input file is already sorted by name. If flag is not set, additional sorting by name will be performed. Only requiredfor paired-end analysis.
+    - name: "--java_memory_size"
+      type: string
+      required: false
+      default: 4G
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/config.vsh.yaml:85]: Remove default

File: src/qualimap/qualimap_rnaseq/script.sh [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+    counts=$(basename "$par_output_counts")
+fi
+
+# disable flags
+[[ "$par_paired" == "false" ]] && unset par_paired
+[[ "$par_sorted" == "false" ]] && unset par_sorted
+
+# Run qualimap
+qualimap rnaseq \
+    --java-mem-size=$par_java_memory_size \
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/script.sh:23]: If you set this to the following, does this work?

```suggestion
    ${meta_memory_mb:+--java-mem-size=${meta_memory_mb}M} \

File: src/qualimap/qualimap_rnaseq/script.sh [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+    --algorithm $par_algorithm \
+    --sequencing-protocol $par_sequencing_protocol \
```
Comment (rcannood) [inline src/qualimap/qualimap_rnaseq/script.sh:25]: Remove default

```suggestion
    ${par_algorithm:+--algorithm $par_algorithm} \
    ${par_sequencing_protocol:+--sequencing-protocol $par_sequencing_protocol} \

File: src/multiqc/script.sh [hunk 23]
```diff
@@ -122,7 +122,6 @@ multiqc \
     "${inputs[@]}" 
 
 # Move outputs
-
```
Comment (rcannood) [inline src/multiqc/script.sh:125]: Undo this change ^^

**Review by rcannood — CHANGES_REQUESTED** [review]
Minor questions and comments

**Review by rcannood — APPROVED** [review]
LGTM!

**Review by rcannood — CHANGES_REQUESTED** [review]
Sorry @dorien-er , I have one last question :bow:

**Review by rcannood — APPROVED** [review]
